File: python/Trees/construct_binary_tree_preorder_inorder_traversal.py
from typing import List, Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        """
        Reconstructs a binary tree from preorder and inorder traversal lists.
        Approach:
        - The first element of preorder is always the root.
        - Find the root in inorder to split left/right subtrees.
        - Recursively build left and right subtrees.
        Time Complexity: O(n^2) (due to list slicing and index search)
        Space Complexity: O(n) (call stack and tree nodes)
        """
        if not preorder or not inorder:
            return None
        root = TreeNode(preorder[0])
        mid = inorder.index(preorder[0])
        root.left = self.buildTree(preorder[1: mid + 1], inorder[:mid])
        root.right = self.buildTree(preorder[mid + 1:], inorder[mid + 1:])
        return root

# -------------------
# Helper Functions for Testing
# -------------------
def tree_to_list(root):
    # Level order traversal to list (None for missing nodes)
    if not root:
        return []
    result = []
    queue = [root]
    while any(queue):
        node = queue.pop(0)
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    # Remove trailing None values
    while result and result[-1] is None:
        result.pop()
    return result

def is_same_tree(p, q):
    if not p and not q:
        return True
    if not p or not q or p.val != q.val:
        return False
    return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)

# -------------------
# Test Cases
# -------------------
def test_build_tree():
    s = Solution()
    # Test case 1: Example
    preorder = [3,9,20,15,7]
    inorder = [9,3,15,20,7]
    root = s.buildTree(preorder, inorder)
    assert tree_to_list(root) == [3,9,20,None,None,15,7]
    # Test case 2: Single node
    preorder = [1]
    inorder = [1]
    root = s.buildTree(preorder, inorder)
    assert tree_to_list(root) == [1]
    # Test case 3: Empty tree
    preorder = []
    inorder = []
    root = s.buildTree(preorder, inorder)
    assert tree_to_list(root) == []
    # Test case 4: Left-skewed tree
    preorder = [1,2,3]
    inorder = [3,2,1]
    root = s.buildTree(preorder, inorder)
    assert tree_to_list(root) == [1,2,None,3]
    # Test case 5: Right-skewed tree
    preorder = [1,2,3]
    inorder = [1,2,3]
    root = s.buildTree(preorder, inorder)
    assert tree_to_list(root) == [1,None,2,None,3]
    print("All test cases passed!")

if __name__ == "__main__":
    test_build_tree()

File: C++/Trees/construct_binary_tree_preorder_inorder_traversal.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    /**
     * Reconstructs binary tree from preorder and inorder traversal.
     *
     * Approach:
     * - First element of preorder is the root.
     * - Find root in inorder to split into left & right subtrees.
     * - Recurse on left and right subtrees.
     *
     * Time Complexity: O(n^2) (finding index in `inorder` at each level)
     * Space Complexity: O(n) (recursion stack + tree nodes)
     */
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty() || inorder.empty()) {
            return nullptr;
        }

        // First element in preorder is always the root
        int rootVal = preorder[0];
        TreeNode* root = new TreeNode(rootVal);

        // Find the index of root in inorder to split left & right subtrees
        auto it = find(inorder.begin(), inorder.end(), rootVal);
        int mid = distance(inorder.begin(), it);

        // Construct vectors for left & right subtrees
        vector<int> leftPre(preorder.begin() + 1, preorder.begin() + 1 + mid);
        vector<int> leftIn(inorder.begin(), inorder.begin() + mid);
        vector<int> rightPre(preorder.begin() + 1 + mid, preorder.end());
        vector<int> rightIn(inorder.begin() + mid + 1, inorder.end());

        root->left = buildTree(leftPre, leftIn);
        root->right = buildTree(rightPre, rightIn);

        return root;
    }
};

/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard headers.

using namespace std;
    ✅ Use standard namespace to avoid typing std:: everywhere.

struct TreeNode { ... }
    ✅ Defines the binary tree node structure.

TreeNode() : val(0), left(nullptr), right(nullptr) {}
    ✅ Default constructor: value = 0, children = nullptr.

TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    ✅ Constructor with value, children = nullptr.

TreeNode(int x, TreeNode *left, TreeNode *right)
    ✅ Constructor with value and pointers to left & right child.

class Solution { ... }
    ✅ Defines the Solution class which contains buildTree.

TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder)
    ✅ Method to build binary tree from preorder & inorder.

if (preorder.empty() || inorder.empty())
    ✅ Base case: if input lists are empty, return nullptr.

int rootVal = preorder[0];
    ✅ The first element of preorder is the root value.

TreeNode* root = new TreeNode(rootVal);
    ✅ Create a new TreeNode with the root value.

auto it = find(inorder.begin(), inorder.end(), rootVal);
    ✅ Find iterator pointing to root in inorder list.

int mid = distance(inorder.begin(), it);
    ✅ Get index (distance from start) of root in inorder list.

vector<int> leftPre(...), leftIn(...), rightPre(...), rightIn(...)
    ✅ Create sub-vectors for left & right subtrees from preorder and inorder.

root->left = buildTree(leftPre, leftIn);
    ✅ Recursively build left subtree.

root->right = buildTree(rightPre, rightIn);
    ✅ Recursively build right subtree.

return root;
    ✅ Return the constructed tree's root.

*/
